Implement Constructor in WebIDL

yjaeseok opened the pull request:
We need to implement constructor in WebIDL.
So I just implement simple constructor.

But we need to implement method overloading mechanism.

I refer to [W3C Spec](https://github.com/w3c/webidl2.js#extended-attributes)
ISSUE=#70

yjaeseok committed:
Implement Constructor in WebIDL

We need to implement constructor in WebIDL.
So I just implement simple constructor.

But we need to implement method overloading mechanism.

I refer to [W3C Spec](https://github.com/w3c/webidl2.js#extended-attributes)
ISSUE=#70

hwanseung commented on generator/idl_parser/idls.ts:
```diff
@@ -29,19 +29,43 @@ export interface InterfaceMember extends TypedIdentifier {
 export const enum DefinitionType {
   Interface,
   Enum,
+  ExtendedAttribute,
```
i think ExtendedAttribute is not same level with Interface or Enum.
(refer to around this lines https://github.com/lunchclass/bacardi/blob/master/generator/idl_parser/idls.ts#L107)
so i think it is better to remove here.
and i am considering to remove DefinitionType enums.
because we can distinguish types from object using |instanceof| keyword without type attributes.

yjaeseok replied:
Okay I try to remove and check other solution .

hwanseung commented on examples/calculator.js:
```diff
@@ -16,7 +16,16 @@
+    new bacardi.Calculator(1000);
+    new bacardi.Calculator(2000);
+    new bacardi.Calculator(3000);
+  });
```
could you add more tests for exception case?
or it is fine to add exception case tests in next patches.

yjaeseok replied:
Actually, I don't know how to use try-catch exception block in Javascript. 
But I will try this one. :)

yjaeseok replied:
@hwanseung I updated it , please check +_+

romandev replied:
Fortunately, we have already other examples to test for try-catch exception.
You can refer to "Test for invalid arguments" test.

yjaeseok committed:
Remove DefinitionType ExtendedAttribute

ExtendedAttribute is not same level with Interface.
So I remove it.

yjaeseok committed:
Modify test case for constructor

Previous test case for constructor is not art.
So I modify and add package chai for exception handling.

With chai package, we can test with expect keyword like below:
```
var badFn = () => {
  new bacardi.Calculator();
};
expect(badFn, TypeError, 'bad arguments');
```

yjaeseok commented:
This is test result !
```shell
./bacardi test

Test for constructor function
(node:9556) Warning: N-API is an experimental feature and could change at any time.
    there is constructor calculator(long createTime)
      ✓ should be create object without error
    there is no constructor calculator()
      ✓ should be failed create object with error
    when we create two objects
      ✓ should be different, two objects

  Test for static function

    ✓ The static function can be called without instantiation.

  isEquals function test for bool type
    ✓ should be equal with expected value

  add function test for long type
    ✓ should be equal with expected value

  sub function test for short type
    ✓ should be equal with expected value

  mul function test for double
    ✓ should be equal with expected value

  div function test for double
    ✓ should be equal with expected value

  Test for invalid arguments
    ✓ Pass invalid numbers
    ✓ No match function signatures


  11 passing (12ms)


```

yjaeseok committed:
Modify test case for constructor

I add to.throw(TypeError) statement.
This statement should be added because the assert works normally.

yjaeseok committed:
Modify test case for constructor(2)

I remove unnecessary arguments expect function.

romandev commented on examples/calculator.js:
```diff
@@ -16,7 +16,31 @@
 const assert = require('assert');
 const bacardi = require('bindings')('bacardi.node');
-const calculator = new bacardi.Calculator();
+const chai = require('chai');
```
We already use Mocha as testing framework. So, you can just use the APIs.
Do you have any reason that we have to add this new testing framework?

yjaeseok replied:
but mocha seems to do not support expect() function.

So I think we cannot handling expect-exception logic test like chai.

romandev reviewed (commented):
Thank you for this patch

yjaeseok committed:
merge remote branch

romandev commented on examples/calculator.js:
```diff
@@ -16,9 +16,33 @@
+      var badFn = () => {
+        new bacardi.Calculator();
+      };
+      expect(badFn).to.throw(TypeError);
```
I thought that we could do it in Mocha as follows:
```Javascript
assert.throws(() => {
     new bacardi.Calculator();
   }, TypeError);
```
If I'm missing something, please correct me :)

yjaeseok replied:
Wow.. looks so simple! I will try this one.

yjaeseok replied:
@romandev great! I changed it like you mention.
Thanks your review!

romandev reviewed (commented):
I left a question.

yjaeseok committed:
Merge remote-tracking branch 'upstream/master'

yjaeseok committed:
remove chai package and modify test case

romandev reviewed (approved):
lgtm
